modules/core/src/smithy4s/http/HttpContractError.scala

```diff
@@ -154,12 +154,19 @@ object MetadataError {
       FailedConstraint.schema.oneOf[MetadataError]("failedConstraint")
     val impossibleDecoding =
       ImpossibleDecoding.schema.oneOf[MetadataError]("impossibleDecoding")
-    union(notFound, wrongType, arityError, failedConstraint) {
-      case n: NotFound           => notFound(n)
-      case w: WrongType          => wrongType(w)
-      case a: ArityError         => arityError(a)
-      case f: FailedConstraint   => failedConstraint(f)
-      case i: ImpossibleDecoding => impossibleDecoding(i)
+
+    union(
+      notFound,
+      wrongType,
+      arityError,
+      failedConstraint,
+      impossibleDecoding
+    ) {
+      case _: NotFound           => 0
+      case _: WrongType          => 1
+      case _: ArityError         => 2
+      case _: FailedConstraint   => 3
+      case _: ImpossibleDecoding => 4
     }
```

Comment:
**thought:** rant time: I have a tiny gripe with the fact that you can't pass Scala types as values without coming up with custom constructs like `Hint.Binding` or `ShapeTag`.

ClassTag doesn't count.

Reply:
for sure

Reply:
really there should be a dual to constructor that verifies that you have one "thing" per case, without having to actually handcraft a pattern-match.